### PR TITLE
Fix: Request for 210 projects returns "Internal server error" (#3886)

### LINKED ISSUE #3886: Request for 210 projects returns "Internal server error"
Request: https://service.azul.data.humancellatlas.org/index/projects?catalog=dcp13&size=210
Result: {"message": "Internal server error"}

From the Data Browser team: "We read all the projects in order to 1) get the diff between catalogs and 2) build a sitemap that lists all projects. We are modifying the differ and sitemap code to use a low size (e.g. 50) and pagination but I wanted to bring this to the team's attention as the large size used to work until just recently".

[edit by @hannes-ucsc]
Proposed solution here: https://github.com/DataBiosphere/azul/issues/3886#issuecomment-1067089108
[/edit]

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: #3886

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `p` tag to titles of partial commits
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>
- [x] All connected issues are resolved partially <sub>or this PR does not have the `partial` label</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only include `p` if the PR is labeled `partial`</sub>
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: src/azul/plugins/__init__.py [hunk 1]
```diff
@@ -331,6 +331,15 @@ def exposed_indices(self) -> Mapping[str, Sorting]:
         """
         raise NotImplementedError
 
+    @property
+    @abstractmethod
+    def page_size_limits(self) -> Mapping[str, int]:
```
Comment (nadove-ucsc) [inline src/azul/plugins/__init__.py]:
```suggestion
    def page_size_limits(self) -> Mapping[EntityType, int]:
```

File: lambdas/service/app.py [hunk 2]
```diff
@@ -509,7 +509,7 @@ def version():
     }
 
 
-def validate_repository_search(params, **validators):
+def validate_repository_search(entity_type, params, **validators):
```
Comment (nadove-ucsc) [inline lambdas/service/app.py]:
```suggestion
def validate_repository_search(entity_type: EntityType, params, **validators):
```

it's a shame there's no type annotations for the other parameters. Maybe just add `Mapping` for `params`?

File: lambdas/service/app.py [hunk 3]
```diff
@@ -546,7 +545,8 @@ def validate_catalog(catalog):
                                 f'Must be one of {set(config.catalogs)}.')
 
 
-def validate_size(size):
+def validate_size(entity_type, size):
```
Comment (nadove-ucsc) [inline lambdas/service/app.py]:
```suggestion
def validate_size(entity_type: EntityType, size):
```

File: src/azul/plugins/__init__.py (hunk 1 above)
Comment (nadove-ucsc) [reply src/azul/plugins/__init__.py]: Same for subclass methods

File: lambdas/service/app.py [hunk 4]
```diff
@@ -902,6 +902,10 @@ def filter_schema(relation: str) -> JSON:
 
 
 def repository_search_params_spec():
+    # Since this spec is used for all entity types and the schema validation
+    # isn't dynamic by which entity type is selected, we have to allow the
+    # largest possible value.
```
Comment (nadove-ucsc) [inline lambdas/service/app.py]: Phrasing is a bit off. How about:

```suggestion
    # This spec is shared across all entity types, so no single value would be correct here.
    # Using the largest possible value ensures that we don't incorrectly reject valid requests, at the cost
    # of accepting some invalid ones.
```

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Nice work, just a few details I think could be improved :+1:

File: src/azul/plugins/__init__.py [hunk 5]
```diff
@@ -331,6 +332,15 @@ def exposed_indices(self) -> Mapping[str, Sorting]:
         """
         raise NotImplementedError
 
+    @property
+    @abstractmethod
+    def page_size_limits(self) -> Mapping[EntityType, int]:
+        """
+        The return value maps an outer entity type of each exposed index to the
+        maximum allowed page size of the corresponding endpoint.
```
Comment (nadove-ucsc) [inline src/azul/plugins/__init__.py]:
```suggestion
        Some types of outer entity aggregates can reach very large sizes, causing timeouts or other errors when
        requests are made to their associated index using large page sizes. This mapping determines the
        maximum page size for each exposed index, allowing us to set lower limits for larger entities while preserving
        the higher limits for smaller ones.
```

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]

**Review by nadove-ucsc — DISMISSED** [review]

File: lambdas/service/app.py [hunk 6]
```diff
@@ -902,6 +905,11 @@ def filter_schema(relation: str) -> JSON:
 
 
 def repository_search_params_spec():
+    # This spec is shared across all entity types, so no single value would be
+    # correct here. Using the largest possible value ensures that we don't
+    # incorrectly reject valid requests, at the cost of accepting some invalid
+    # ones.
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Don't understand this comment. The specs aren't used for validation, only documentation. You can specify here whatever you want without affecting what's accepted and what's rejected.

If you have to pick a value, use the value that works everywhere.

Also consider not specifying a maximum value at all.

You could also enforce a minimal maximum pages size i.e. a page size that must be supported in every catalog and for every entity type.

File: src/azul/plugins/__init__.py [hunk 7]
```diff
@@ -331,6 +332,19 @@ def exposed_indices(self) -> Mapping[str, Sorting]:
...
+    @property
+    @abstractmethod
+    def page_size_limits(self) -> Mapping[EntityType, int]:
+        """
+        Some types of outer entity aggregates can reach very large sizes,
+        causing timeouts or other errors when requests are made to their
+        associated index using large page sizes. This mapping determines the
+        maximum page size for each exposed index, allowing us to set lower
+        limits for larger entities while preserving the higher limits for
+        smaller ones.
```
Comment (hannes-ucsc) [inline src/azul/plugins/__init__.py]: A bit verbose. Entity sizes differ between catalogs and entity types, so it's kind of obvious why one would scale the page size limits with them.

File: lambdas/service/app.py [hunk 8]
```diff
@@ -902,6 +905,11 @@ def filter_schema(relation: str) -> JSON:
...
+    max_page_size = max(app.metadata_plugin.page_size_limits.values())
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: This using the default catalog and ignores the others. But I think we have this problem elsewhere. If you end up leaving this approach in, add a FIXME referring to https://github.com/DataBiosphere/azul/issues/4288.

But see my comment above.

Comment (hannes-ucsc) [reply lambdas/service/app.py]: But most importantly, you have to document the quirks to the Azul *user*, not to the *developer*. The right place for that is the description of the actual parameter.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: lambdas/service/app.py (hunk 6 above)
Comment (dsotirho-ucsc) [reply lambdas/service/app.py]: The Swagger UI uses the values specified in the spec to do some client side validation.

swagger-validation

The maximum can be omitted.

Screen Shot 2023-10-09 at 4 59 30 PM

File: src/azul/plugins/__init__.py (hunk 5 above)
Comment (hannes-ucsc) [inline src/azul/plugins/__init__.py]:
```suggestion
        The return value maps the entity type of each exposed index to the
        maximum page size allowed by the corresponding endpoint.
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The fact that the maximum page size depends on catalog and entity type, needs to be documented so that *users* of the API are made aware of this.

File: src/azul/plugins/__init__.py [hunk 9]
```diff
@@ -331,6 +332,15 @@ def exposed_indices(self) -> Mapping[str, Sorting]:
...
         raise NotImplementedError
 
+    @property
+    @abstractmethod
+    def page_size_limits(self) -> Mapping[EntityType, int]:
+        """
+        The return value maps the entity type of each exposed index to the
+        maximum page size allowed by the corresponding endpoint.
+        """
+        raise NotImplementedError
```
Comment (hannes-ucsc) [inline src/azul/plugins/__init__.py]: We should integrate this with `exposed_indexes`:

```diff
Subject: [PATCH] WIP
---
Index: src/azul/plugins/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/__init__.py b/src/azul/plugins/__init__.py
--- a/src/azul/plugins/__init__.py	(revision 35a8fd70d69a78d03a92796fd41c26c0fdc3b0fa)
+++ b/src/azul/plugins/__init__.py	(date 1697506061739)
@@ -123,6 +123,7 @@
 class Sorting:
     field_name: FieldName
     descending: bool = attr.ib(default=False)
+    max_page_size: int = 1000
 
     @property
     def order(self) -> str:
@@ -327,8 +328,8 @@
         """
         The indices for which the service provides an `/index/…` endpoint.
         The return value maps the outer entity type of each exposed index to the
-        default values of the request parameters that control the ordering of
-        hits returned by the corresponding endpoint.
+        default values of the request parameters that control the paging and
+        ordering of hits returned by the corresponding endpoint.
         """
         raise NotImplementedError
 
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: lambdas/service/app.py [hunk 10]
```diff
@@ -975,8 +979,9 @@ def repository_search_params_spec():
         ),
         params.query(
             'size',
-            schema.optional(schema.with_default(10, type_=schema.in_range(min_page_size, max_page_size))),
-            description='The number of hits included per page.'),
+            schema.optional(schema.with_default(10, type_=schema.in_range(min_page_size, None))),
+            description='The number of hits included per page. '
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Should use `fd` here, and fix the siblings, too.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**